lib/file-source/src/file_watcher.rs

```diff
@@ -247,12 +251,26 @@ fn null_reader() -> impl BufRead {
     io::Cursor::new(Vec::new())
 }
 
-// Tweak of https://github.com/rust-lang/rust/blob/bf843eb9c2d48a80a5992a5d60858e27269f9575/src/libstd/io/mod.rs#L1471
-// After more than max_size bytes are read as part of a single line, this discard the remaining bytes
-// in that line, and then starts again on the next line.
+/// Read up to `max_size` bytes from `reader`, splitting by `delim`
+///
+/// The function reads up to `max_size` bytes from `reader`, splitting the input
+/// by `delim`. If a `delim` is not found in `reader` before `max_size` bytes
+/// are read then the reader is polled until `delim` is found and the results
+/// are discarded. Else, the result is written into `buf`.
+///
+/// The return is unusual. In the Err case this function has not written into
+/// `buf` and the caller should not examine its contents. In the Ok case if the
+/// inner value is None the caller should retry the call as the buffering read
+/// hit the end of the buffer but did not find a `delim` yet, indicating that
+/// we've sheered a write or that there were no bytes available in the `reader`
+/// and the `reader` was very sure about it. If the inner value is Some the
+/// interior `usize` is the number of bytes written into `buf`.
```

Comment:
There's more context on why it's in this unusual state in the description of #4089. Very much a classic "we'll clean this up soon" 🙂

Reply:
A good friend of mine was prone to say "there's nothing permanent like temporary". :joy: